models/dsn_helpers.go

```diff
@@ -0,0 +1,151 @@
+
+// DSNforMySQL generates MySQL DSN string from service and exporter.
+func DSNforMySQL(service *Service, exporter *Agent, db string) string {
+	// TODO TLSConfig: "true", https://jira.percona.com/browse/PMM-1727
+	// TODO Other parameters?
+
+	cfg := mysql.NewConfig()
+	cfg.User = pointer.GetString(exporter.Username)
+	cfg.Passwd = pointer.GetString(exporter.Password)
+	cfg.Net = "tcp"
+	host := pointer.GetString(service.Address)
+	port := pointer.GetUint16(service.Port)
+	cfg.Addr = net.JoinHostPort(host, strconv.Itoa(int(port)))
+	cfg.Timeout = 1 * time.Second
+	cfg.DBName = db
+
+	// QAN code in pmm-agent uses reform which requires those fields
+	cfg.ClientFoundRows = true
+	cfg.ParseTime = true
+
+	return cfg.FormatDSN()
+}
+
+// DSNforPostgreSQL generates Postgres DSN string from service and exporter.
+func DSNforPostgreSQL(service *Service, exporter *Agent) string {
+	q := make(url.Values)
+	q.Set("sslmode", "disable") // TODO: make it configurable
+	q.Set("connect_timeout", "1")
+
+	host := pointer.GetString(service.Address)
+	port := pointer.GetUint16(service.Port)
+	username := pointer.GetString(exporter.Username)
+	password := pointer.GetString(exporter.Password)
+
+	u := &url.URL{
+		Scheme:   "postgres",
+		Host:     net.JoinHostPort(host, strconv.Itoa(int(port))),
+		Path:     "postgres",
+		RawQuery: q.Encode(),
+	}
+	switch {
+	case password != "":
+		u.User = url.UserPassword(username, password)
+	case username != "":
+		u.User = url.User(username)
+	}
+
+	return u.String()
+}
+
+// DSNforMongoDB generates MongoDB DSN string from service and exporter.
+func DSNforMongoDB(service *Service, exporter *Agent) string {
+	host := pointer.GetString(service.Address)
+	port := pointer.GetUint16(service.Port)
+	username := pointer.GetString(exporter.Username)
+	password := pointer.GetString(exporter.Password)
+
+	u := &url.URL{
+		Scheme: "mongodb",
+		Host:   net.JoinHostPort(host, strconv.Itoa(int(port))),
+	}
+	switch {
+	case password != "":
+		u.User = url.UserPassword(username, password)
+	case username != "":
+		u.User = url.User(username)
+	}
+
+	return u.String()
+}
+
+// FindDSNByServiceIDandPMMAgentID resolves DSN by service id.
+func FindDSNByServiceIDandPMMAgentID(q *reform.Querier, serviceID, pmmAgentID, db string) (string, error) {
+	var result string
+
+	svc, err := FindServiceByID(q, serviceID)
+	if err != nil {
+		return "", err
+	}
+
+	var agentType AgentType
+	switch svc.ServiceType {
+	case MySQLServiceType:
+		agentType = MySQLdExporterType
+	case MongoDBServiceType:
+		agentType = MongoDBExporterType
+	case PostgreSQLServiceType:
+		agentType = PostgresExporterType
+	default:
+		return "", status.Errorf(codes.FailedPrecondition, "Couldn't resolve dsn, as service is unsupported")
+	}
+
+	exporters, err := FindAgentsByServiceIDAndAgentType(q, serviceID, agentType)
+	if err != nil {
+		return "", err
+	}
+
+	fexp := make([]*Agent, 0)
+	for _, e := range exporters {
+		if pointer.GetString(e.PMMAgentID) == pmmAgentID {
```

Comment:
minor comment: I think we can add it to sql query FindAgentsByServiceIDAndAgentType.

Reply:
Then it will be `FindAgentsByServiceIDAndAgentTypeAndPmmAgentID`